### PR TITLE
upgrade nemo road to integrate the viewing manifest directly

### PR DESCRIPTION
**Pull request by Corentinfaye**

### CODE REVIEW

File: templates/main/multipassage.html [hunk 1]
```diff
@@ -114,7 +126,7 @@ <h3 class="entry-title text-center">{{object.collections.current.title}}<a id="{
             {%set all_reffs = objects|map(attribute='subreference')|list %}
             {%set new_reffs = objects|map(attribute='subreference')|list %}
             {% set navigation = nav(text, all_texts, new_reffs, loop.index0) %}
-            {% if current_user.project_team == False and 'andecavensis' in text.objectId %}
+            {% if current_user.project_team == False %}
```
Comment (sonofmun) [inline templates/main/multipassage.html]: The check of whether 'andecavensis' is in the objectId needs to stay. Why did you delete it in the first place?

File: formulae/nemo.py [hunk 2]
```diff
@@ -539,20 +540,44 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
         #ids contient tous les differents objectId
         ids = objectIds.split('+')
         translations = {}
+        view=1
         for i in ids:
+            if "manifest" in i:
+                i = i[9:]
```
[the remark this answers is not in the document]
Comment (Corentinfaye) [reply formulae/nemo.py]: I think Regex is better because we can read what we decide to delete on this string

File: templates/main/multipassage.html [hunk 3]
```diff
@@ -35,14 +45,14 @@ <h3 class="entry-title text-center">{{object.collections.current.title}}<a id="{
 
 
 {% macro version_menu(objId, inRefs) %}
-
 {% if objects %}
```
Comment (sonofmun) [inline templates/main/multipassage.html]: We should be able to remove these lines here as well since they are now valid for the whole template.

File: formulae/include/python3.6m [hunk 4]
```diff
@@ -0,0 +1 @@
+/usr/include/python3.6m
```
Comment (sonofmun) [inline formulae/include/python3.6m]: Should this file be included? If so, what is its purpose? If not, then remove it from your branch and add it to the `.gitignore` file.

File: formulae/nemo.py [hunk 5]
```diff
@@ -459,6 +460,7 @@ def r_passage(self, objectId, subreference, lang=None):
         collection = self.get_collection(objectId)
         if isinstance(collection, CtsWorkMetadata):
             editions = [t for t in collection.children.values() if isinstance(t, CtsEditionMetadata)]
+            print(editions)
```
Comment (sonofmun) [inline formulae/nemo.py]: You should remove this print statement.

File: formulae/nemo.py [hunk 6]
```diff
@@ -539,25 +541,42 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
         #ids contient tous les differents objectId
         ids = objectIds.split('+')
         translations = {}
+        view=1
```
Comment (sonofmun) [inline formulae/nemo.py]: Put spaces around the `=`.

File: formulae/nemo.py [hunk 7]
```diff
@@ -539,25 +541,42 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
...
         for i in ids:
+            if "manifest" in i:
+                i = re.sub(r'^manifest:', '', i)
             p = self.resolver.getMetadata(self.resolver.getMetadata(i).parent.id)
             translations[i] = [m for m in p.readableDescendants if m.id not in ids]
         passage_data = {'template': 'main::multipassage.html', 'objects': [], "translation": translations}
         subrefers = subreferences.split('+')
         for i, id in enumerate(ids):
             if self.check_project_team() is True or id in self.open_texts:
+                v = False
```
Comment (sonofmun) [inline formulae/nemo.py]: Lines 554-557 should be moved before line 553.

File: formulae/nemo.py [hunk 8]
```diff
@@ -539,25 +541,42 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
...
-                if 'previous_search' in session:
-                    result_sents = [x['sents'] for x in session['previous_search'] if x['id'] == id]
-                    if result_sents:
-                        d['text_passage'] = self.highlight_found_sents(d['text_passage'],
-                                                                       self.convert_result_sents(result_sents))
+                if v == True:
+                    formulae = self.app.picture_file["manifest:"+id]
+                    d["objectId"] = "manifest:"+id
+                    d["div_v"] = "manifest"+str(view)
+                    view = view+1
```
Comment (sonofmun) [inline formulae/nemo.py]: For lines 565-568, always put spaces around the `+`. This is the best practice in Python to always have spaces around mathematical operators (+, -, /, *, =, etc.).

File: formulae/nemo.py [hunk 9]
```diff
@@ -539,25 +541,42 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
...
                     subref = subrefers[i]
                 d = self.r_passage(id, subref, lang=lang)
                 del d['template']
-                d["IIIFviewer"] = id in self.app.picture_file
-                if 'previous_search' in session:
-                    result_sents = [x['sents'] for x in session['previous_search'] if x['id'] == id]
-                    if result_sents:
-                        d['text_passage'] = self.highlight_found_sents(d['text_passage'],
-                                                                       self.convert_result_sents(result_sents))
+                if v == True:
```
Comment (sonofmun) [inline formulae/nemo.py]: Since `v` is a boolean, you can just write `if v:`.

File: formulae/nemo.py [hunk 10]
```diff
@@ -539,25 +541,42 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
...
+                    formulae = self.app.picture_file["manifest:"+id]
+                    d["objectId"] = "manifest:"+id
+                    d["div_v"] = "manifest"+str(view)
+                    view = view+1
+                    del d['text_passage']
+                    #this viewer work when the library or archiv give an IIIF API for the external usage of theirs books
+                    d["manifest"] = url_for('viewer.static', filename=formulae["manifest"])
+                    d["title"] = formulae["title"]
+                else:
+                    d["IIIFviewer"] = "manifest:"+id in self.app.picture_file
```
Comment (sonofmun) [inline formulae/nemo.py]: Again, add spaces around your `+`.

File: formulae/viewer/routes.py [hunk 11]
```diff
@@ -6,88 +6,13 @@
...
-        return current_app.config['nemo_app'].render(template='main::index.html', url=dict())
-'''
 
 
 
 @bp.route("/<objectId>", methods=["GET"])
-def addviewer(objectId):
+def fullscreenviewer(objectId):
+    if "manifest" in objectId:
+        objectId=objectId[9:]
```
Comment (sonofmun) [inline formulae/viewer/routes.py]: Use the regular expression here instead of the index.

File: formulae/viewer/routes.py [hunk 12]
```diff
@@ -96,112 +21,14 @@ def addviewer(objectId):
         objectId = str(editions[0].id)
     if current_app.config['nemo_app'].check_project_team() is True or objectId in current_app.config['nemo_app'].open_texts:
         template = {'manifest': 'viewer::miradorviewer.html'}
-        formulae = current_app.picture_file[objectId]
-        passage_data = ''
-        if request.args.get('embedded', False):
-            template = {'manifest': "viewer::multiviewermirador.html"}
-            passage_data = current_app.config['nemo_app'].r_passage(objectId, '1')
+        formulae = current_app.picture_file['manifest:'+objectId]
```
Comment (sonofmun) [inline formulae/viewer/routes.py]: Add spaces around the `+`.

File: formulae/viewer/routes.py [hunk 13]
```diff
@@ -96,112 +21,14 @@ def addviewer(objectId):
...
-            #create the link if it's an image in local server or take the complet link
-            if "/images/formulae-1" in str(images[view]):
-                link_picture=current_app.IIIFserver +str(images[view])
-            else:
-                link_picture=str(images[view])
-            return current_app.config['nemo_app'].render(template=template['local'], picture=link_picture, objectId=objectId,
-                                                         current_view=view, total_views=len(images), text=passage_data,
-                                                         url=dict(), town=town, codex=codex, current_folios=current_folios, folios=folios)
+        return current_app.config['nemo_app'].render(template=template['manifest'], manifest=manifest
+                                                     ,objectId=objectId, url=dict())
```
Comment (sonofmun) [inline formulae/viewer/routes.py]: The comma before `objectId=objectId` should be on the previous line.

File: templates/main/multipassage.html (hunk 3 above)
Comment (sonofmun) [reply templates/main/multipassage.html]: It doesn't look like you removed these lines. Does the code not work if these lines are removed?

File: tests/test_routes.py [hunk 14]
```diff
@@ -252,21 +252,21 @@ def test_authorized_project_member(self):
...
             self.assertIn('Angers 3' + _(' hat keine Edition.'), r.get_data(as_text=True))
-            c.get('/viewer/urn:cts:formulae:andecavensis.form005.lat001', follow_redirects=True)
+            r = c.get('/viewer/urn:cts:formulae:andecavensis.form001', follow_redirects=True)
+            self.assertTemplateUsed('viewer::miradorviewer.html')
+            r = c.get('/texts/urn:cts:formulae:andecavensis.form002.lat001+manifest:urn:cts:formulae:andecavensis.form002.lat001/passage/1+1', follow_redirects=True)
+            self.assertTemplateUsed('main::multipassage.html')
+            r = c.get('texts/manifest:urn:cts:formulae:andecavensis.form002.lat001+manifest:urn:cts:formulae:andecavensis.form001.lat001/passage/1+1', follow_redirects=True)
+            self.assertTemplateUsed('main::multipassage.html')
+            r = c.get('/texts/urn:cts:formulae:andecavensis.form002.lat001+manifest:urn:cts:formulae:andecavensis.form002.lat001/passage/1+all', follow_redirects=True)
+            self.assertTemplateUsed('main::multipassage.html')
```
Comment (sonofmun) [inline tests/test_routes.py]: What are the differences between the last three tests that all call `multipassage.html`? If it is only the identifiers that are being called, then you only need a single test.

Comment (Corentinfaye) [reply tests/test_routes.py]: It was useful before the merge of the two branch of r_passage

File: formulae/nemo.py [hunk 15]
```diff
@@ -12,6 +12,7 @@
 from lxml import etree
 from .errors.handlers import e_internal_error, e_not_found_error, e_unknown_collection_error
 import re
+import os
```
Comment (sonofmun) [inline formulae/nemo.py]: Are you still using the `os` package here? If not, you should delete the import.

File: formulae/nemo.py [hunk 16]
```diff
@@ -539,25 +540,43 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
...
         for i in ids:
+            if "manifest" in i:
+                i = re.sub(r'^manifest:', '', i)
             p = self.resolver.getMetadata(self.resolver.getMetadata(i).parent.id)
             translations[i] = [m for m in p.readableDescendants if m.id not in ids]
         passage_data = {'template': 'main::multipassage.html', 'objects': [], "translation": translations}
         subrefers = subreferences.split('+')
         for i, id in enumerate(ids):
+            v = False
+            m = re.match(r'.*(?=\.)',id)
```
Comment (sonofmun) [inline formulae/nemo.py]: There should be a space after the comma in the `re.match` statement.

File: formulae/nemo.py (hunk 9 above)
Comment (sonofmun) [reply formulae/nemo.py]: You didn't change this line.

File: formulae/nemo.py [hunk 17]
```diff
@@ -539,25 +540,43 @@ def r_multipassage(self, objectIds, subreferences, lang=None):
...
-                d["IIIFviewer"] = id in self.app.picture_file
-                if 'previous_search' in session:
-                    result_sents = [x['sents'] for x in session['previous_search'] if x['id'] == id]
-                    if result_sents:
-                        d['text_passage'] = self.highlight_found_sents(d['text_passage'],
-                                                                       self.convert_result_sents(result_sents))
+                if v == True:
+                    formulae = self.app.picture_file[m.group(0)]
+                    d["objectId"] = "manifest:" + id
+                    d["div_v"] = "manifest"+str(view)
```
Comment (sonofmun) [inline formulae/nemo.py]: There should be a space before and after the `+`.

File: formulae/viewer/routes.py [hunk 18]
```diff
@@ -96,112 +20,14 @@ def addviewer(objectId):
         objectId = str(editions[0].id)
     if current_app.config['nemo_app'].check_project_team() is True or objectId in current_app.config['nemo_app'].open_texts:
         template = {'manifest': 'viewer::miradorviewer.html'}
-        formulae = current_app.picture_file[objectId]
-        passage_data = ''
-        if request.args.get('embedded', False):
-            template = {'manifest': "viewer::multiviewermirador.html"}
-            passage_data = current_app.config['nemo_app'].r_passage(objectId, '1')
+        formulae = current_app.picture_file['manifest:' + re.match(r'.*(?=\.)',objectId).group(0)]
```
Comment (sonofmun) [inline formulae/viewer/routes.py]: In the `re.match` statement, there should be a space between the comma and `objectId`.
